### ISSUE TITLE
Logi Pop Keys emoji keys not supported

### ISSUE THREAD

**Issue by Brli** [pwr-Solaar/Solaar#1372]
**Information**

- Solaar version 1.1.0
- Distribution:
- Kernel version 5.15.7.arch1
- Output of `solaar show`:

<details>

```
Bolt Receiver
  Device path  : /dev/hidraw2
  USB id       : 046d:C548
  Serial       : 43353536463033374537393838313845
  Has 2 paired device(s) out of a maximum of 6.
  Notifications: wireless, software present (0x000900)
  Device activity counters: 1=154, 2=102

  1: Logi POP Keys
     Device path  : None
     WPID         : B365
     Codename     : Logi POP Keys
     Kind         : keyboard
     Protocol     : HID++ 4.5
     Serial number: F0C214D3
     Model ID:      B36500000000
     Unit ID:       F0C214D3
        Bootloader: BL1 44.00.B0005
          Firmware: RBK 69.00.B0005
             Other: 
     The power switch is located on the (unknown).
     Supports 31 HID++ 2.0 features:
         0: ROOT                   {0000}   
         1: FEATURE SET            {0001}   
         2: DEVICE FW VERSION      {0003}   
            Firmware: Bootloader BL1 44.00.B0005 B36567A459DC01
            Firmware: Firmware RBK 69.00.B0005 B36567A459DC01
            Firmware: Other   
            Unit ID: F0C214D3  Model ID: B36500000000  Transport IDs: {'btleid': 'B365'}
         3: DEVICE NAME            {0005}   
            Name: Logi POP Keys
            Kind: keyboard
         4: WIRELESS DEVICE STATUS {1D4B}   
         5: RESET                  {0020}   
         6: DEVICE FRIENDLY NAME   {0007}   
            Friendly Name: Logi POP Keys
         7: UNIFIED BATTERY        {1004}   
         8: REPROG CONTROLS V4     {1B04}   
         9: CHANGE HOST            {1814}   
            更改主機號碼        : 1:cjslappy
        10: HOSTS INFO             {1815}   
            Host 0 (paired): cjslappy
            Host 1 (unpaired): 
            Host 2 (unpaired): 
        11: K375S FN INVERSION     {40A3}   
            互換 Fx 鍵功能 (saved): False
            互換 Fx 鍵功能        : False
        12: LOCK KEY STATE         {4220}   
        13: KEYBOARD DISABLE KEYS  {4521}   
            Disable keys (saved): {'1': False, '16': False, '8': False}
            Disable keys        : {'1': False, '8': False, '16': False}
        14: MULTIPLATFORM          {4531}   
            Set OS (saved): 0
            Set OS        : Windows
        15: KEYBOARD LAYOUT 2      {4540}   
        16: unknown:00C3           {00C3}   
        17: DEVICE RESET           {1802}   internal, hidden, unknown:000010
        18: unknown:1803           {1803}   internal, hidden, unknown:000010
        19: CONFIG DEVICE PROPS    {1806}   internal, hidden, unknown:000010
        20: unknown:1816           {1816}   internal, hidden, unknown:000010
        21: OOBSTATE               {1805}   internal, hidden
        22: unknown:1830           {1830}   internal, hidden, unknown:000010
        23: unknown:1891           {1891}   internal, hidden, unknown:000008
        24: unknown:18A1           {18A1}   internal, hidden, unknown:000010
        25: unknown:1E00           {1E00}   hidden
        26: unknown:1E02           {1E02}   internal, hidden
        27: unknown:1602           {1602}   
        28: unknown:1EB0           {1EB0}   internal, hidden, unknown:000010
        29: unknown:1861           {1861}   internal, hidden, unknown:000010
        30: unknown:18B0           {18B0}   internal, hidden, unknown:000010
     Has 20 reprogrammable keys:
         0: Host Switch Channel 1     , default: HostSwitch Channel 1        => HostSwitch Channel 1      
             is FN, FN sensitive, analytics key events, pos:1, group:0, group mask:empty
             reporting: default
         1: Host Switch Channel 2     , default: HostSwitch Channel 2        => HostSwitch Channel 2      
             is FN, FN sensitive, analytics key events, pos:2, group:0, group mask:empty
             reporting: default
         2: Host Switch Channel 3     , default: HostSwitch Channel 3        => HostSwitch Channel 3      
             is FN, FN sensitive, analytics key events, pos:3, group:0, group mask:empty
             reporting: default
         3: Show Desktop              , default: Show Desktop                => Show Desktop              
             is FN, FN sensitive, reprogrammable, divertable, analytics key events, pos:4, group:0, group mask:empty
             reporting: default
         4: unknown:010A              , default: unknown:00DF                => unknown:00DF              
             is FN, FN sensitive, reprogrammable, divertable, analytics key events, pos:5, group:0, group mask:empty
             reporting: default
         5: unknown:011C              , default: unknown:00F1                => unknown:00F1              
             is FN, FN sensitive, reprogrammable, divertable, analytics key events, pos:6, group:0, group mask:empty
             reporting: default
         6: Previous Fn               , default: Previous                    => Previous                  
             is FN, FN sensitive, reprogrammable, divertable, analytics key events, pos:7, group:0, group mask:empty
             reporting: default
         7: Play/Pause Fn             , default: Play/Pause                  => Play/Pause                
             is FN, FN sensitive, reprogrammable, divertable, analytics key events, pos:8, group:0, group mask:empty
             reporting: default
         8: Next Fn                   , default: Next                        => Next                      
             is FN, FN sensitive, reprogrammable, divertable, analytics key events, pos:9, group:0, group mask:empty
             reporting: default
         9: Mute Fn                   , default: Mute                        => Mute                      
             is FN, FN sensitive, reprogrammable, divertable, analytics key events, pos:10, group:0, group mask:empty
             reporting: default
        10: Volume Down Fn            , default: Volume Down                 => Volume Down               
             is FN, FN sensitive, reprogrammable, divertable, analytics key events, pos:11, group:0, group mask:empty
             reporting: default
        11: Volume Up Fn              , default: Volume Up                   => Volume Up                 
             is FN, FN sensitive, reprogrammable, divertable, analytics key events, pos:12, group:0, group mask:empty
             reporting: default
        12: unknown:0103              , default: unknown:00D8                => unknown:00D8              
             nonstandard, reprogrammable, divertable, analytics key events, pos:0, group:0, group mask:empty
             reporting: default
        13: unknown:0104              , default: unknown:00D9                => unknown:00D9              
             nonstandard, reprogrammable, divertable, analytics key events, pos:0, group:0, group mask:empty
             reporting: default
        14: unknown:0105              , default: unknown:00DA                => unknown:00DA              
             nonstandard, reprogrammable, divertable, analytics key events, pos:0, group:0, group mask:empty
             reporting: default
        15: unknown:0106              , default: unknown:00DB                => unknown:00DB              
             nonstandard, reprogrammable, divertable, analytics key events, pos:0, group:0, group mask:empty
             reporting: default
        16: unknown:0107              , default: unknown:00DC                => unknown:00DC              
             nonstandard, reprogrammable, divertable, analytics key events, pos:0, group:0, group mask:empty
             reporting: default
        17: unknown:0108              , default: LedToggle                   => LedToggle                 
             nonstandard, reprogrammable, divertable, analytics key events, pos:0, group:0, group mask:empty
             reporting: default
        18: F Lock                    , default: Do Nothing One              => Do Nothing One            
             is FN, analytics key events, pos:0, group:0, group mask:empty
             reporting: default
        19: unknown:0034              , default: Do Nothing One              => Do Nothing One            
             nonstandard, analytics key events, pos:0, group:0, group mask:empty
             reporting: default
     Battery: 100%, discharging.

  2: Logi POP Mouse
     Device path  : None
     WPID         : B030
     Codename     : Logi POP Mous
     Kind         : mouse
     Protocol     : HID++ 4.5
     Serial number: D235B5D6
     Model ID:      B03000000000
     Unit ID:       D235B5D6
        Bootloader: BL1 55.00.B0004
          Firmware: RBM 20.00.B0004
             Other: 
     The power switch is located on the (unknown).
     Supports 32 HID++ 2.0 features:
         0: ROOT                   {0000}   
         1: FEATURE SET            {0001}   
         2: DEVICE FW VERSION      {0003}   
            Firmware: Bootloader BL1 55.00.B0004 B030668A051E
            Firmware: Firmware RBM 20.00.B0004 B030668A051E
            Firmware: Other   
            Unit ID: D235B5D6  Model ID: B03000000000  Transport IDs: {'btleid': 'B030'}
         3: DEVICE NAME            {0005}   
            Name: Logi POP Mouse
            Kind: mouse
         4: WIRELESS DEVICE STATUS {1D4B}   
         5: RESET                  {0020}   
         6: CRYPTO ID              {0021}   
         7: DEVICE FRIENDLY NAME   {0007}   
            Friendly Name: Logi POP Mouse
         8: UNIFIED BATTERY        {1004}   
         9: REPROG CONTROLS V4     {1B04}   
            Key/Button Actions (saved): {'264': 264, '82': 82}
            Key/Button Actions        : {'82': 82, '264': 264}
        10: CHANGE HOST            {1814}   
            更改主機號碼        : 1:cjslappy
        11: HOSTS INFO             {1815}   
            Host 0 (paired): cjslappy
            Host 1 (unpaired): 
            Host 2 (unpaired): 
        12: XY STATS               {2250}   
        13: LOWRES WHEEL           {2130}   
            Wheel Reports: HID
            Scroll Wheel Diversion (saved): False
            Scroll Wheel Diversion        : False
        14: ADJUSTABLE DPI         {2201}   
            靈敏度 (DPI) (saved): 1600
            靈敏度 (DPI)        : 1600
        15: unknown:00C3           {00C3}   
        16: DEVICE RESET           {1802}   internal, hidden, unknown:000010
        17: unknown:1803           {1803}   internal, hidden, unknown:000010
        18: CONFIG DEVICE PROPS    {1806}   internal, hidden, unknown:000010
        19: unknown:1816           {1816}   internal, hidden, unknown:000010
        20: OOBSTATE               {1805}   internal, hidden
        21: unknown:1830           {1830}   internal, hidden, unknown:000010
        22: unknown:1891           {1891}   internal, hidden, unknown:000008
        23: unknown:18A1           {18A1}   internal, hidden, unknown:000010
        24: unknown:1E00           {1E00}   hidden
        25: unknown:1E02           {1E02}   internal, hidden
        26: unknown:1E22           {1E22}   internal, hidden, unknown:000010
        27: unknown:1602           {1602}   
        28: unknown:1EB0           {1EB0}   internal, hidden, unknown:000010
        29: unknown:1861           {1861}   internal, hidden, unknown:000010
        30: unknown:18B1           {18B1}   internal, hidden, unknown:000010
        31: unknown:920A           {920A}   internal, hidden
     Has 5 reprogrammable keys:
         0: Left Button               , default: Left Click                  => Left Click                
             mse, analytics key events, pos:0, group:1, group mask:empty
             reporting: default
         1: Right Button              , default: Right Click                 => Right Click               
             mse, analytics key events, pos:0, group:1, group mask:empty
             reporting: default
         2: Middle Button             , default: Mouse Middle Button         => Mouse Middle Button       
             mse, reprogrammable, divertable, raw XY, analytics key events, pos:0, group:2, group mask:g1,g2
             reporting: default
         3: unknown:0108              , default: LedToggle                   => LedToggle                 
             mse, reprogrammable, divertable, raw XY, analytics key events, pos:0, group:2, group mask:g1,g2
             reporting: default
         4: Virtual Gesture Button    , default: Virtual Gesture Button      => Virtual Gesture Button    
             divertable, virtual, raw XY, force raw XY, pos:0, group:3, group mask:empty
             reporting: default
     Battery: 100%, discharging.


```
</details>

**Describe the bug**

**To Reproduce**
1. pair the keyboard with Logi Bolt
2. start a input area (console, notepad app, browser, anything)
3. press any of the four button
4. nothing shows

all the tools revealed in [Arch Wiki](https://wiki.archlinux.org/title/Keyboard_input) report nothing.

1. showkey --scancodes dies out after 10 seconds as if nothing input, even though I constantly go through the four keys
2. evtest and xev doesn't return anything.

**Additional context**

The keymap is set to en, 104 generic keyboard

under Windows, the keys (all four of them) are reported as keycode 231 under https://keycode.info, and is output directly as the preset emoji.

I don't know if it is a driver issue, or xkb configuration issue though.

Let me know if you need more test, or has some idea to at least get the keystroke signal.

**Comment by pfps** [comment]
Solaar is not an input driver.  So if the keyboard produces regular input that is not correctly handled by the Linux keyboard input drivers and other parts of Linux, Solaar does not directly intervene.   
What Solaar can do in some cases for keyboards is to change what the keyboard does when a key is pressed.  To see what might be possible requires the output ot `solaar show`, which is missing here.

**Comment by Brli** [comment]
solaar show done

**Comment by pfps** [comment]
Many Logitech keyboards like the emoji keys the Pop has do things that only work in Windows and maybe only then with the Logitech driver.  
Solaar can modify the action of some of these keys to something useful, but this requires that Solaar knows about the actions, which does not appear to be the case for many of the Pop keys.  Solaar can also modify the action of some of these keys to send a special message that Solaar can then use to trigger its rules.
To make this work well, Solaar needs to know what the numbers in the output of `solaar show` mean, e.g.,
```
13: unknown:0104              , default: unknown:00D9    
```
You can help by following the process in issue #1307 to find out what, for example, key 0104 actually is.
Solaar from this repository already does know about  some of the keys because they are shared with the MX Keys for business, and MX Mini.

The bottom line is that you probably can get these keys to do something useful, but maybe not actually insert the emoji.

**Comment by pfps** [comment]
PR #1453 has new keys, based on information from Logitech.  Please try it out and see whether your keyboard's keys are covered.


To clone and use Solar from its GitHub repository
```
git clone https://github.com/pwr-Solaar/Solaar.git
cd Solaar
```
Run Solaar as bin/solaar from this directory.


To run PR #1453, first clone Solaar if you have not already done so and cd to the clone directory.  The first time you download the pull request, fetch it into a new branch and checkout that branch, as in:
```
git fetch origin pull/1453/head:pull_1453
git checkout pull_1453
```
To download a new version of the pull request, fetch it and then set your pull branch
to the new fetch, as in:
```
git checkout pull_1453
git fetch origin pull/1453/head
git reset --hard FETCH_HEAD
```

**Comment by Brli** [comment]
so, yes, the configuration window, but xev still receive nothing, and nothing is shown when pressing thos emoji keys under terminal or any notepad apps (kate, vscode etc.)
![Screenshot_20220218_100717](https://user-images.githubusercontent.com/4042566/154604010-9a8744ed-a0ea-4a13-bcd3-df361d925d9f.png)
n

**Comment by pfps** [comment]
Solaar can't directly change how the Linux input driver handles these keys when they generate normal input events (or don't generate any input events).  Solaar also can't directly change how the keys work, for the Logi Pop keyboard.  What Solaar can do is "divert" these keys so that they produce notifications that can trigger Solaar rules.  The Solaar rules are then responsible for calling programs that do something or simulating input that does something. 
To make the emoji keys insert emojis you would have to write Solaar rules that trigger on the Key and do something that is recognized as a request to insert an emoji.  I'm not sure what that something is.

**Comment by Brli** [comment]
I've set them to diverted, but, no default rules related are shown in the
edit rul UI

also, since xev doesn't popup events, I suppose solar can do little thing
on it.

I'll check to u in few days.

Btw, the issue link u gave me about debugging isn't helpful, or maybe it's
due to I didn't digest the thread the right way...

Peter F. Patel-Schneider ***@***.***> 於 2022年2月18日 週五 下午12:51
寫道：

**Comment by pfps** [comment]
Do you mean that no rules are shown at all?   If you are running Wayland, Solaar rules do not work at the moment.  Try running with `solaar -dd` and see if there are warning messages.

**Comment by pfps** [comment]
As PR #1453 adds the keys, I'm going to merge it and close this issue.  Note that Solaar can't fix how the Linux drivers handle the keys, all that can be done is to divert the keys and write Solaar rules that process the resulting notifications.

If you are having problems with rules feel free to open a new issue on them.

**Comment by avi-ran** [comment]
For those wondering how to set these rules for the emojis to work with Logitech PopKeys on Ubuntu, here is what I've done :

I were able to make all 4 emojis (from top to bottom) working by selecting "diverted" ([https://ibb.co/GnWCybX](https://ibb.co/GnWCybX)) for each emoji keys in the main Solaar panel and I added 4 rules (one for each button) like so (sorry it's a french screenshot): [https://ibb.co/6m1kWsm](https://ibb.co/6m1kWsm)

😍🥲😄❤

**Comment by Brli** [comment]
hi, @avi-ran, would you mind sharing your `~/.config/solaar/rules.yaml` here?

**Comment by cbeley** [comment]
Has anyone found a way to divert the "Open Emoji Panel" button differently depending on whether it is coming from the keyboard or the mouse? Unfortunately, both devices seem to emit it as the same code and I'm not seeing a way to test for a specific originating device in the rule editor.

**Comment by pfps** [comment]
Solaar doesn't have a rule condition for that at the moment.   Please file an enhancement issue if you want this to be added to Solaar.

**Comment by cbeley** [comment]
Thanks! I was worried I was just missing something. I wrote up https://github.com/pwr-Solaar/Solaar/issues/1995 for anyone else who wants to follow along.

I'll also say that diverting of all the special keys otherwise works perfectly for me.

**Comment by nosrednawall** [comment]
Thanks
